modules/ray_tracing/src/utils/RayTracingAngularQuadrature.C

```diff
@@ -0,0 +1,276 @@
+  {
+    x[i] = x_copy[sorted_indices[i]];
+    w[i] = w_copy[sorted_indices[i]];
+  }
+}
+
+void
+RayTracingAngularQuadrature::rotate(const libMesh::Point & rotation_direction)
+{
+  _current_rotation_direction = rotation_direction;
+
+  libMesh::DenseMatrix<Real> rotation_matrix;
+  rotationMatrix(rotation_direction.unit(), rotation_matrix);
+
+  _current_directions.clear();
+  _current_directions.reserve(_phi.size());
+
+  _current_weights.clear();
+  _current_weights.reserve(_w.size());
+
+  _current_polar_sins.clear();
+  _current_polar_sins.reserve(_polar_sin.size());
+
+  libMesh::DenseVector<Real> omega(3);
+  libMesh::DenseVector<Real> omega_p(3);
+  Point direction;
+
+  for (std::size_t q = 0; q < _phi.size(); ++q)
+  {
+    // Get direction as a DenseVector for rotation
+    omega(0) = sqrt(1 - _mu[q] * _mu[q]) * cos(_phi[q]);
+    omega(1) = sqrt(1 - _mu[q] * _mu[q]) * sin(_phi[q]);
+    omega(2) = _mu[q];
+
+    // Rotate and create the direction
+    rotation_matrix.vector_mult(omega_p, omega);
+    direction(0) = omega_p(0);
+    direction(1) = omega_p(1);
+    direction(2) = omega_p(2);
+
+    std::size_t l = _current_directions.size();
+    // If 2D, project to the xy plane and see if any other
+    // projected directions are a duplicate
+    if (_dim == 2)
+    {
+      direction(2) = 0;
+      direction /= direction.norm();
+
+      for (l = 0; l < _current_directions.size(); ++l)
+        if (_current_directions[l].absolute_fuzzy_equals(direction))
+          break;
+    }
+
+    // If l is the current size, it means a duplicate was not found
+    // so we can add the direction and initialize the weight/sins
+    if (l == _current_directions.size())
+    {
+      _current_directions.push_back(direction);
+      _current_weights.resize(l + 1);
+      _current_polar_sins.resize(l + 1);
+    }
+
+    // Add to weights and sins for this direction
+    _current_weights[l].push_back(_w[q]);
+    _current_polar_sins[l].push_back(_polar_sin[q]);
+  }
+}
+
+const libMesh::Point &
+RayTracingAngularQuadrature::getDirection(const unsigned int l) const
+{
+  mooseAssert(hasDirection(l), "Does not have direction");
+  return _current_directions[l];
+}
+
+const std::vector<Real> &
+RayTracingAngularQuadrature::getWeights(const unsigned int l) const
+{
+  mooseAssert(hasDirection(l), "Does not have direction");
+  return _current_weights[l];
+}
+
+Real
+RayTracingAngularQuadrature::getTotalWeight(const unsigned int l) const
+{
+  mooseAssert(hasDirection(l), "Does not have direction");
+  return std::accumulate(_current_weights[l].begin(), _current_weights[l].end(), (Real)0);
+}
+
+const std::vector<Real> &
+RayTracingAngularQuadrature::getPolarSins(const unsigned int l) const
+{
+  mooseAssert(hasDirection(l), "Does not have direction");
+  return _current_polar_sins[l];
+}
+
+std::size_t
+RayTracingAngularQuadrature::numPolar(const unsigned int l) const
+{
+  mooseAssert(hasDirection(l), "Does not have direction");
+  if (_dim == 3)
```

Comment:
Wait, how does this work? In opt mode the mooseAssert line is removed so that the if () then refers to the next line, i.e. the return statement?